File: P258712.py
def solution(friends, gifts):
    answer = 0
    dic = {}
    score = {}
    answer_list = {}
    for friend in friends:
        dic[friend] = {}
        score[friend] = 0
        answer_list[friend] = 0
        for friend_ in friends:
            if friend != friend_:
                dic[friend][friend_] = 0
                
    for gift in gifts:
        give, take = map(str, gift.split())
        dic[give][take] += 1
        score[give] += 1
        score[take] -= 1
    
    for x in friends:
        for y in friends:
            if x != y:
                if dic[x][y] == dic[y][x]:
                    if score[x] == score[y]:
                        continue
                    elif score[x] > score[y]:
                        answer_list[x] += 1
                
                elif dic[x][y] > dic[y][x]:
                    answer_list[x] += 1
                    
    for friend in friends:
        if answer_list[friend] > answer:
            answer = answer_list[friend]
            
    return answer

File: P258712.java
import java.util.HashMap;

public class P258712 {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        HashMap<String, HashMap<String, Integer>> dic = new HashMap<>();
        HashMap<String, Integer> score = new HashMap<>();
        HashMap<String, Integer> answer_list = new HashMap<>();
        for (String friend : friends) {
            dic.put(friend, new HashMap<>());
            score.put(friend, 0);
            answer_list.put(friend, 0);

            for (String friend_ : friends) {
                if (!friend.equals(friend_)) dic.get(friend).put(friend_, 0);
            }
        }

        for (String gift : gifts) {
            String[] splitedGift = gift.split(" ");
            String give = splitedGift[0];
            String take = splitedGift[1];
            dic.get(give).put(take, dic.get(give).get(take)+1);
            score.put(give, score.get(give)+1);
            score.put(take, score.get(take)-1);
        }

        for (String x : friends) {
            for (String y : friends) {
                if (!x.equals(y)) {
                    if (dic.get(x).get(y) == dic.get(y).get(x).intValue()) {
                        if (score.get(x) == score.get(y).intValue()) {
                        } else if (score.get(x) > score.get(y)) answer_list.put(x, answer_list.get(x)+1);
                    }

                    else if (dic.get(x).get(y) > dic.get(y).get(x)) answer_list.put(x, answer_list.get(x)+1);
                }
            }
        }

        for (String friend : friends) {
            if (answer_list.get(friend) > answer) answer = answer_list.get(friend);
        }
        return answer;
    }
}
